### PR TITLE
Preflight missing multipass

### PR DESCRIPTION
**Pull request by evandandrea**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] If this is a bugfix. Have you checked that there is a bug report open for the issue you are trying to fix on [bug reports](https://bugs.launchpad.net/snapcraft)?
- [ ] If this is a new feature. Have you discussed the design on the [forum](https://forum.snapcraft.io)?
- [x] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

Fixes https://bugs.launchpad.net/snapcraft/+bug/1792374

Feedback welcome on structure. @sergiusens suggested asking to install multipass further up, but it's not clear to me where. A decorator on the lifecycle functions?

### CODE REVIEW

File: snapcraft/project/_sanity_checks.py [hunk 1]
```diff
@@ -32,7 +35,21 @@
...
+def conduct_build_environment_sanity_check(provider: str):
+    if provider == "multipass":
+        _check_multipass_installed()
+
+
+def _check_multipass_installed():
+    if not shutil.which("multipass"):
+        if sys.platform == "linux":
+            if not shutil.which("snap"):
+                raise errors.SnapMissingLinuxError()
```
Comment (sergiusens) [inline snapcraft/project/_sanity_checks.py]: nice covering this, we will always run from the snap moving forward, but this makes it easy to report on checks when running from source.

File: snapcraft/project/_sanity_checks.py [hunk 2]
```diff
@@ -32,7 +35,21 @@
...
 
-# TODO: Add conduct_environment_sanity_check() to be run in the build environment
+
+def conduct_build_environment_sanity_check(provider: str):
+    if provider == "multipass":
+        _check_multipass_installed()
+
+
+def _check_multipass_installed():
+    if not shutil.which("multipass"):
```
Comment (sergiusens) [inline snapcraft/project/_sanity_checks.py]: how about
```python
if shutil.which("multipass"):
    return

if sys.platform != "linux":
    raise errors.MultipassMissingNonLinuxError()

if shutil.which("snap"):
   raise errors.MultipassMissingLinuxError()
else:
    raise errors.SnapMissingLinuxError()
```

Which makes it a little less nested and easier to add future checks in as well.

File: snapcraft/cli/lifecycle.py [hunk 3]
```diff
@@ -15,7 +15,7 @@
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
 import typing
-
+import sys
```
[the remark this answers is not in the document]
Comment (evandandrea) [reply snapcraft/cli/lifecycle.py]: …it is below `import typing`, which is also core. Am I misreading?

**Comment by sergiusens** [comment]
Job is green https://travis-ci.org/snapcore/snapcraft/builds/460246673?utm_source=github_status&utm_medium=notification

**Comment by evandandrea** [comment]
I'm +1 on the copy.

**Comment by evandandrea** [comment]
Works here:
```
$ snapcraft
You need multipass installed to build snaps which use the base keyword.
Would you like to install it now? [y/N]: y
multipass (beta) 2018.10.1 from Canonical✓ installed
Channel latest/beta for multipass is closed; temporarily forwarding to beta.
Waiting for multipass...
Launching a VM.
failed to launch: CPU does not support KVM extensions.
An error occurred when trying to launch the instance with 'multipass': returned exit code 2.
Ensure that 'multipass' is setup correctly and try again.
ubuntu@subtly-needed-chaffinch:~/offlineimap$ sudo snap remove multipass
multipass removed
ubuntu@subtly-needed-chaffinch:~/offlineimap$ snapcraft
You need multipass installed to build snaps which use the base keyword.
Would you like to install it now? [y/N]:
multipass is required to continue.
```

**Comment by sergiusens** [comment]
unrelated failure

**Comment by sergiusens** [comment]
Confirmed working
```
sergiusens@angmar:~/projects/test$ snapcraft 
You need multipass installed to build snaps which use the base keyword.
Would you like to install it now? [y/N]: y
multipass (beta) 2018.10.1 from Canonical✓ installed
Channel latest/beta for multipass is closed; temporarily forwarding to beta.
Waiting for multipass...
Launching a VM.
Launched: snapcraft-my-snap-name                                                                                          
2018-11-28T04:01:00Z INFO Waiting for restart...
core 16-2.36.1 from 'canonical' installed
snapcraft 3.0+git43.g7fde5ba installed
core18 18 from Canonical✓ installed
Hit:1 http://archive.ubuntu.com/ubuntu bionic InRelease                                                                   
Hit:2 http://archive.ubuntu.com/ubuntu bionic-updates InRelease                                                           
Hit:3 http://archive.ubuntu.com/ubuntu bionic-backports InRelease                                                         
Hit:4 http://security.ubuntu.com/ubuntu bionic-security InRelease                                                         
Reading package lists... Done                                                                                             
Building dependency tree                                                                                                  
Reading state information... Done                                                                                         
11 packages can be upgraded. Run 'apt list --upgradable' to see them.                                                     
Pulling my-part                                                                                                           
Building my-part                                                                                                          
Staging my-part                                                                                                           
Priming my-part                                                                                                           
Snapping 'my-snap-name' |                                                                                                 
Snapped my-snap-name_0.1_amd64.snap 
```

**Pull request merged**